### ISSUE TITLE
[feature request] easier custom ordering of legend entries

### ISSUE THREAD

**Issue by HDembinski** [matplotlib/matplotlib#8283]
Summary
=======

Currently, when you plot several artists on the same axis and use the pyplot.legend(), the ordering of the entries is somewhat mysterious for the average user, see http://stackoverflow.com/questions/22263807/how-is-order-of-items-in-matplotlib-legend-determined. 

Like many of my peers, I use matplotlib to make plots for science publications and regularly want to customize the order of the items in the legend when I polish plots. I argue that this is a common thing to do, and therefore it should be easy to do with the API.

My request is to add a new keyword `lorder` to artist constructors, similar to `zorder`. While `zorder` handles the drawing order, `lorder` handles the order in which items appear in the legend. 

Current state
==========

tacaswell explained on SO that the order of legend items is an implementation detail and that's fine. There are ways to re-order the items, but they are somewhat cumbersome. One recommended way is to use 
```python
[...]
handles, labels = axis.get_legend_handles_labels()
```
and then reorder to your hearts desire. This is ok when you want to reorder based on a rule, but it is not convenient when you want to re-order manually (because you first have to find out what the current order is, unpack the tuples, reorder them, repack them).

The other recommended way is to keep the return values of pyplot
```python
[...]
art1, = pyplot.plot(...)
art2, = pyplot.plot(...)
# I always forget how to do it for errorbar
# art3, = pyplot.errorbar(...) # fails, unpacking works differently here
plt.legend((art1, art2), ("label1", "label2"))
```
As you can see, I always have trouble remembering how to unpack the return value correctly for stuff other than pyplot.plot(...) so that it works with plt.legend(...). Surely, other users must feel the same way. I would say that this method relies on knowing more about the matplotlib internals than the average user wants to know.

Proposed solution
=============
The user simply specifies the new `lorder` keyword in calls to pyplot
```python
[...]
pyplot.plot(..., lorder=2, label="second") # appears second in legend
pyplot.plot(..., lorder=3, label="third") # appears third in legend
pyplot.errorbar(..., lorder=1, label="first") # appears first in legend
plt.legend()
```
This code is rather clear and makes it very easy to specify the ordering of items in the legend. As far as I understand tacaswell, it should be easy to handle internally as well, just make `lorder` part of the internal sorting criterion.

A more explicit alternative is keyword `labelorder` instead of the shorter `lorder`.

**Comment by story645** [comment]
I wonder if there's a way to pass in the preferred sorting as an argument to legend so that there's only 1 place the api changes instead of in a million places? maybe by passing in the labels?

**Comment by tacaswell** [comment]
I am not convinced that learning another keyword is easier that learning the return types from plotting methods.  Keeping track of what artists the plotting methods return is (in my view) the best way.  It is clear, gives the user as much control as they want, and does not rely on any behind-the-scene conventions or magic.

If we want to support this, adding it to `Artist` would let it propagate through from kwargs everywhere without much code change.

How would ties be broken under `'lorder'` and what would the default be?

Would it be simpler to sort the legend by zorder?  We already have that concept, but we might not want to mix the two.

One other concern about this is that for thing that we automatically find, this is great, but as soon as the user needs to make a legend entry for things _not_ via the automatic they are going to have to understand artists anyway and now `lorder` is only used if the user implements it them selves.  I think there is some value in making the learning curve a bit steeper in places to avoid cliffs further on ("ah, you want to do something that looks _slightly_ different, here learn a new API" is no good).

**Comment by anntzer** [comment]
I am strongly -1 on this proposal, for the reasons mentioned by @tacaswell (it's basically a kludge that doesn't even generalize well).  Adding an `order` kwarg as suggested by @story645 is only going to raise more issues (what do you do when there are duplicates in the label list?).

I am not convinced that mixing with zorder is going to help at all either...

**Comment by story645** [comment]
Hmm, as much as I'm like no more examples, then what about solving this in the docs somehow/somewhere?

**Comment by NelleV** [comment]
I would just extend the legend_demo example.

**Comment by story645** [comment]
I'm 👍 on @NelleV's suggestion.

**Comment by efimovvl** [comment]
Would implementing this feature as suggested by @HDembinski conflict with the existing way of creating the ordering in the legend box? 

I understand that that you wish to encourage proper use of creating the legend for the graph by using the object that plotting functions return. However, for people who are new and want to do a quick legend with desired order for their graph items (like errorbar), I think this feature would be useful. Provided it does not conflict with the way it is currently done.

My group and I wish to implement this feature and create a pull request on completion. It would be encouraging to know that the pull request will be considered, even if majority of contributers do not think it's a good feature to have.

**Comment by efiring** [comment]
In agreement with other comments above, I am not in favor of adding yet another artist attribute (lorder) or of mixing this with zorder.

Here is an alternative approach, offered with trepidation because it is adding another kwarg to what is perhaps the single most kwarg-overladen and overly-complex object in mpl, the legend:

Add a kwarg, e.g., `legend(use_labels=['dogs', 'cats'])`.

Then the legend code would look for the artists with those labels, and use them with their associated handles (the artists such as line objects), in that order.  It would include in the legend *only* artists with matching labels.  I think this would be quite easy to implement, and would be both readable and easy to use.

**Comment by HDembinski** [comment]
I am happy to see the discussion and I am also glad for the support of @efimovvl. I would have tried to implement this myself if there was enough interest, perhaps we can collaborate on a PR?

However, there were some good questions raised whether this is a good idea at all, or whether there is a better way to get the same functionality. Let me address these points.

__Why is it easier to learn yet another keyword rather than the return types of plotting methods?__ The main point is that `lorder` just requires me to remember one keyword, which has a _universal_ effect on any plotting method. The problem with the return types is that unpacking them works differently, depending on the plotting method. The behavior is _not uniform_ and there is more information for me to remember.

__What to do in case of ties?__
Nothing special, the order is then an implementation detail. It is a user fails to provide an unambiguous `lorder`, it is their problem (just like with `zorder`).

__Sort by zorder?__
Different responsibilities should not be mixed.

__Isn't the current approach a nice bridge from simple to more advanced use of matplotlib?__
I don't think so. I think we all agree that simple and common things should be simple to do with the API. I do all my scientific plotting with matplotlib and I use it as my only tool for scientific plotting for many years now. Please trust me that reordering legend items is a common thing you want to do when you polish plots for publications. If it is a common thing to do, the API should provide a simple way of doing it.

__Why a keyword for the Artist and not for the `legend(...)` call?__
I think this is not a good solution, because it reduces readability and introduces non-locality of string literals. Let me elaborate. In this alternative suggested API, I need to specify the labels twice in different places, once in the calls to the plotting methods and in the legend call.
```python
pyplot.plot(..., label="complex long label1")
pyplot.plot(..., label="complex long label2")
[...] # my be some large piece between
pyplot.legend(use_labels=["complex long label1", "complex long label2"])
```
Repeating the labels manually is error prone and having redundant literals in the code is never elegant. Information in code should be local and not redundant. There is also a practical aspect. In the scientific context, the labels are usually not short strings like "cats" and "dogs", but rather long LaTeX code. Here is a real-life example
```$\mathrm{Var}\left(\frac{\Delta \hat E}{\sigma_\mathrm{res}(\Delta \hat E)}\right)$```
You don't want to repeat that in your code. Now you could put the labels in a tuple
```python
labels = ("complex label1", "complex label2")
pyplot.plot(..., label=labels[0])
pyplot.plot(..., label=labels[1])
[...]
pyplot.legend(use_labels=labels)
```
That would work, but it is less readable. I like how the label literals are put directly in the plotting calls. This approach nicely bundles all information (what is plotted and how it is labeled) together in one place. That's a great and very useful design. I think it greatly helps with the readability of the code, so we should try to keep that.

Summary:
I still believe that `lorder` is the nicest solution. Users of matplotlib are used to tons of keywords which tweak all kinds of behavior. That's just how the matplotlib API works and that's ok. You get used to it and then you start to expect things to work this way. Learning-curve wise, it is most important that things are done consistently. Or to say it with the "Zen of Python":

"There should be one - and preferably only one - obvious way to do it."

In matplotlib, this means when I want to customize the defaults that matplotlib choses for me, I specify a keyword locally in a plotting call which changes these defaults. My proposal follows this design approach. Information stays bundled together locally.

**Comment by u55** [comment]
May I offer an alternative choice that requires no new kwargs? I don't mind unpacking the artists that are returned by `plot`, `errorbar`, etc., but I dislike the need to specify both the artist sequence and label sequence to `legend` if I want to control the legend order. It would be nice if `legend` would automatically pickup the label for each artist if it is specified in the call to `plot`, `errorbar`, etc. Here is an example:

```python
p0, = plt.plot(..., label="complex long label1")
p1, = plt.plot(..., label="complex long label2")

# This picks up the labels that I specified, but I have no control over the legend ordering.
plt.legend()
# Currently this makes default labels named "Line2D(0)", "Line2D(1)", etc.,
# rather than picking up "complex long label1", etc.
plt.legend([p1, p0])
# If I want to do this reordering, I have to also specify the label sequence, 
# which seems superfluous since I already specified the labels for each artist.
plt.legend([p1, p0], ["complex long label2", "complex long label1"])
```
I propose to change the API such that calling `legend` with an artist sequence argument but no label sequence argument will use the labels specified in each `plot` call, etc., and only fall back to "Line2D(0)", etc. if an artist label has not already been specified.

In other words, the 1-arg call to legend will use the labels previously attached to each artist and provide a default fallback; and the 2-arg call to legend will override the labels regardless of whether a label was already attached to each artist.

I would prefer to be able to make this shortened call to `legend` because I like attaching each label to the artist in the call to `plot`, where I can clearly read that the label matches the data, and I am always worried that I will misapply the labels to the wrong artists if I try to manually sort both the artist sequence and label sequence that are currently needed in the call to `legend`.

**Comment by u55** [comment]
### Update

Upon reading the source code of `axes.legend`, I discovered that the functionality that I wanted is already available by using the undocumented kwarg `handles`, i.e.:

```python
plt.legend(handles=[p1, p0]) # This works!
```
It turns out, using the 1-arg call to legend is actually equivalent to using the undocumented kwarg `labels`:
```python
plt.legend(labels=[p1, p0]) # same as plt.legend([p1, p0])
```
which I suppose calls `str` on each artist, which must be where the "Line2D(0)" label names come from. This is totally counterintuitive because in python, usually the last positional arguments are optional, not the first!

With this new information, I propose to document  how the positional argument to `legend` is interpreted if `len(args)==1`, and document the `handles` and `labels` kwarg in the docstring of `axes.legend`, and update the legend example demo to utilize the `handles` kwarg as a simpler way to control the ordering than using the `get_legend_handles_labels` function.

**Comment by efimovvl** [comment]
@u55 I think the bigger concern is that the Axes items may return different types. Not necessarily a handle item (handle item is an Artist or Container).
For example: 
a = Axis.plot(...) returns a list of 2Dlines (subclass of Artist).
b = Axis.errorbar(...) returns a Container instance (which is basically a tuple for multiple Artist objects)

If you call 
plt.legend(labels=[a, b])
it will complain that 'a' is a list and doesn't have the 'get_label' function.

Thus currently it is the case that for each item you want to plot, you need to figure out how to un-wrap it to get the handle object.

**Comment by HDembinski** [comment]
@efimovvl: Thank you for wrapping it up in your own words, yes, that is exactly my point of concern! :)
@u55: I think the functionality you found in the source code is also documented here: http://matplotlib.org/users/legend_guide.html

Design-wise, I think there are two ways to go. My main concerns are that the code is
 * expressive: as the Zen of Python says, "Readability counts"
 * uniform: implementation details should not leak into the user-side of the API and force the user to understand and remember these details (I argue that the structure of return values of pyplot.plot(...) and pyplot.errorbar(...) are implementation details)
 * local: information that belongs together has to be nearby in the code.

These two solutions follow these requirements:

plotting-call focus
=============
My original proposal, legend drawing is controlled by keywords in plotting calls. The content of the legend and the order of the legend entries is specified in the plotting call, not in the legend call.

legend-call focus
=============
The legend call *internally* unwraps the return value of plotting calls uniformly.
```python
a = plt.plot(...)
b = plt.errorbar(...)
plt.legend((a, b), ("complex long label1", "complex long label2"))
```
From the user POV, ```a``` and ```b``` represent the graphical objects drawn by the respective plotting functions, so this is an intuitive interface and the implementation detail is hidden of what the return value actually is. It is probably not too difficult to implement the unwrapping in the legend call.

**Comment by story645** [comment]
I think the proposed solution of writing unwrappers that are called in legend is fairly reasonable. It addresses the problem and doesn't add to/change any api.

**Comment by HDembinski** [comment]
I can try to work on a PR if the proposal is acceptable.

**Comment by anntzer** [comment]
This is a much more reasonable proposal, but, as always, the devil is in the details.  What do you plan to do with
```
ls = plt.plot(x1, y1, x2, y2) # `ls` is a tuple of 2 lines
hs = plt.hist(np.random.rand(3, 3)) # hs is a tuple of the various return values of `hist`
plt.legend((ls, hs), (...))
```
?

**Comment by HDembinski** [comment]
@anntzer crap, you are right. that basically kills proposal 2 and we are back to proposal 1, as far as I can see.

**Comment by efimovvl** [comment]
With a group of colleagues we have implemented this feature, the way it was first described. I finally made the PR for it, see #8507

**Comment by u55** [comment]
I have another proposal which would simplify my use case 99% of the time. Instead of having the legend order be an "implementation detail", why not have the default legend ordering be the same as the ordering of the calls to `plot`, `errorbar`, `hist`, `scatter`, etc., regardless of the compound artist return type? This is usually what I want anyway. I think this would also satisfy the principle of least astonishment for new users. Then, if I still want to change the default ordering, I am happy to use the more powerful API methods that are already available (i.e. the 'handles' and 'labels' kwargs of `legend`).

**Comment by HDembinski** [comment]
@u55 It thought of that as well. It would solve most of my issue as well, but it is not elegant. It couples two things which are (in principle) unrelated: call order and legend order.

The system should not force me to rearrange the call order in my script to produce the order in which I want the legend entries to appear.

This may make sense as long as you use pyplot to write sequential scripts, but imagine you build a plotting framework based on the object-oriented API of matplotlib. Then it becomes less predictable which calls come before others.

Finally, it goes against the Python principle "Explicit is better than implicit".

**Comment by u55** [comment]
@HDembinski I agree that matching the legend order to the artist call order fits better within the "state machine" API of `pyplot`, but even when using the object-oriented API it seems like a more reasonable default than the current "implementation detail", which to the user, might as well be random. This would already be sufficient for day-to-day scripts. However, for the more complicated "unpredictable plotting framework" that you describe, it makes sense to me that you would keep track of the artist handles yourself in a list and pass them _"explicitly"_ to `legend`.

**Comment by anntzer** [comment]
I think it is clear at this point that none of the core developers consider this proposal to be a good idea, so I am closing the issue and the associated PR.  Of course, if another core dev would like to champion this (or a similar proposal), he should feel free to reopen this.
Note that #8389 seems to have more traction among core devs, and would essentially solve the issue from a different angle.
